swayfmt/src/comments.rs

```diff
@@ -0,0 +1,66 @@
+use std::fmt::Write;
+use sway_ast::token::Comment;
+use sway_types::Spanned;
+
+use crate::{formatter::FormattedCode, Formatter, FormatterError};
+
+// Given a start, an end and a CommentMap, return references to all comments
+// contained within a given start and end of a span, in an exclusive range.
+pub fn get_comments_between(start: usize, end: usize, formatter: &mut Formatter) -> Vec<&Comment> {
+    let mut comments = vec![];
+    let iter = formatter.comment_map.clone().into_keys();
+
+    for bs in iter {
+        if bs.start > start && bs.end < end {
+            if let Some(comment) = formatter.comment_map.get(&bs) {
+                comments.push(comment)
+            }
+        }
+    }
+
+    comments
+}
+
+// Given a start, an end and a CommentMap, removes and returns all comments in a
+// CommentMap contained within a given start and end of a span, in an exclusive range.
+pub fn take_comments_between(start: usize, end: usize, formatter: &mut Formatter) -> Vec<Comment> {
+    let mut comments = vec![];
+    let iter = formatter.comment_map.clone().into_keys();
+
+    for bs in iter {
+        if bs.start > start && bs.end < end {
+            if let Some(comment) = formatter.comment_map.remove(&bs) {
+                comments.push(comment)
+            }
+        }
+    }
+
+    comments
+}
```

Comment:
`BTreeMap` does not have an equivalent for `Range` that also removes the yielded elements, however it does have a [`drain_filter` method](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.drain_filter).

A more efficient (performance and code-wise) alternative to this function using `drain_filter` might look like:

```rust
formatter.comment_map.drain_filter(|bs, _v| bs.start >= start && bs.end <= end)
```

Reply:
`drain_filter` seems to be unstable atm - we can consider adding this when it's stabilized. Will create a good first issue for this

Reply:
Ahh apologies, I missed the nightly label!